src/transformers/retrieval_rag.py

```diff
@@ -0,0 +1,535 @@
+
+        Args:
+            question_hidden_states (:obj:`torch.Tensor` of shape :obj:`(batch_size, vector_size)`:
+                A batch of query vectors to retrieve with.
+            n_docs (:obj:`int`):
+                The number of docs retrieved per query.
+
+        Ouput:
+            total_scores (:obj:`torch.Tensor` of shape :obj:`(batch_size, n_docs)`
+                The retrieval scores of the retrieved docs per query.
+            total_examples (:obj:`List[dict]`):
+                The retrieved examples per query.
+        """
+
+        # non-ddp initialization (init_retrieval() is called at ddp initialization, if no ddp, then it's never called,
+        # so it has to be initalized separately.
+        if not dist.is_initialized() and not self.retriever.is_initialized():
+            logger.info("Initializing index at first query")
+            self.retriever.init_index()
+
+        # single GPU training
+        if not dist.is_initialized():
+            doc_ids, retrieved_doc_embeds = self._main_retrieve(question_hidden_states, n_docs)
+            return retrieved_doc_embeds, self.retriever.get_doc_dicts(doc_ids)
+
+        # distributed training
+        world_size = dist.get_world_size(group=self.process_group)
+
+        # gather logic
+        gather_list = None
+        if self._is_main():
+            gather_list = [torch.empty(question_hidden_states.shape, dtype=torch.float32) for _ in range(world_size)]
+        dist.gather(question_hidden_states, dst=0, gather_list=gather_list, group=self.process_group)
+
+        # scatter logic
+        n_queries = question_hidden_states.shape[0]
+        scatter_ids = []
+        scatter_vectors = []
+        if self._is_main():
+            assert len(gather_list) == world_size
+            ids, vectors = self._main_retrieve(torch.cat(gather_list), n_docs)
+            scatter_ids = self._chunk_tensor(ids, n_queries)
+            scatter_vectors = self._chunk_tensor(vectors, n_queries)
+        doc_ids = self._scattered(scatter_ids, [n_queries, n_docs], target_type=torch.int64)
+        retrieved_doc_embeds = self._scattered(scatter_vectors, [n_queries, n_docs, question_hidden_states.shape[1]])
+
+        return retrieved_doc_embeds, self.retriever.get_doc_dicts(doc_ids)
+
+    def __call__(self, question_input_ids, question_hidden_states, prefix, n_docs=None):
+        # TODO(patrick, quention): would be nice to make retriever framework independent => we could pass `return_tensors='pt'" here
+        n_docs = n_docs if n_docs is not None else self.n_docs
+        retrieved_doc_embeds, docs = self.retrieve(question_hidden_states, n_docs)
+
+        input_strings = self.question_encoder_tokenizer.batch_decode(question_input_ids, skip_special_tokens=True)
+        context_input_ids, context_attention_mask = self.postprocess_docs(docs, input_strings, prefix, n_docs)
+
+        return context_input_ids, context_attention_mask, retrieved_doc_embeds
```

Comment:
We should have a self-documented dataclass or BatchEncoding output here like the models/tokenizers.

Since this object is supposed to be mostly framework independent in the future. I think outputting a dict (or copying the BatchEncoding method with its casting methods but without the token specific alignement methods) is probably the safest options (ModelOuputs are still not crazy TF compatible I feel like).